File: web/src/api/passthrough.ts
import { http } from '@/lib/http.ts';

// Import, enumeration, and relay setup can outlive a normal request timeout.
const startTimeout = 120 * 1000;
const listTimeout = 15 * 1000;

// The imported device as the exporter described it.
export type PassthroughDevice = {
  busId: string;
  idVendor: string;
  idProduct: string;
  speed: string;
  class: number;
};

export type PassthroughStatus = {
  active: boolean;
  mode: PassthroughMode;
  exporter: string;
  udc: string;
  port: number;
  hub: string;
  bus: number;
  address: number;
  pid: number;
  hidSurrendered: boolean;
  startedAt: string;
  device?: PassthroughDevice | null;
};

export type PassthroughMode = 'hybrid' | 'exact';

// One entry of the exporter's device list. unsupported names the class that
// makes the device isochronous, and is absent when nothing about it needs the
// override.
export type PassthroughRemoteDevice = {
  busId: string;
  idVendor: string;
  idProduct: string;
  speed: string;
  class: number;
  unsupported?: string;
};

export type PassthroughDeviceList = {
  devices: PassthroughRemoteDevice[];
};

// the current session, or a zero status when none is running
export function getPassthrough() {
  return http.get('/api/vm/passthrough');
}

// Import busId from exporter and start the selected relay. allowIsochronous
// lifts the refusal that keeps streaming devices out of an Exact session.
export function startPassthrough(
  exporter: string,
  busId: string,
  mode: PassthroughMode,
  allowIsochronous = false
) {
  return http.post(
    '/api/vm/passthrough/start',
    { exporter, busId, mode, allowIsochronous },
    { timeout: startTimeout }
  );
}

// Enumerate the exporter's devices so a busid does not have to be typed.
export function listPassthroughDevices(exporter: string) {
  return http.post('/api/vm/passthrough/devices', { exporter }, { timeout: listTimeout });
}

// Stop the relay, detach the port, and restore the gadget.
export function stopPassthrough() {
  return http.post('/api/vm/passthrough/stop');
}

File: server/proto/passthrough.go
package proto

import "time"

// The imported device as the exporter described it. Both modes refuse
// isochronous endpoints.
type PassthroughDevice struct {
	BusID     string `json:"busId"`
	IDVendor  string `json:"idVendor"`
	IDProduct string `json:"idProduct"`
	Speed     string `json:"speed"`
	Class     uint8  `json:"class"`
}

// HIDSurrendered distinguishes Exact from Hybrid sessions.
type GetPassthroughRsp struct {
	Active         bool               `json:"active"`
	Mode           string             `json:"mode"`
	Exporter       string             `json:"exporter"`
	UDC            string             `json:"udc"`
	Port           uint32             `json:"port"`
	Hub            string             `json:"hub"`
	Bus            uint32             `json:"bus"`
	Address        uint32             `json:"address"`
	Pid            int                `json:"pid"`
	HIDSurrendered bool               `json:"hidSurrendered"`
	StartedAt      time.Time          `json:"startedAt"`
	Device         *PassthroughDevice `json:"device"`
}

// One entry of the exporter's device list. Unsupported carries why the device
// cannot be relayed and is empty when it can.
type PassthroughRemoteDevice struct {
	BusID       string `json:"busId"`
	IDVendor    string `json:"idVendor"`
	IDProduct   string `json:"idProduct"`
	Speed       string `json:"speed"`
	Class       uint8  `json:"class"`
	Unsupported string `json:"unsupported,omitempty"`
}

type ListPassthroughRsp struct {
	Devices []PassthroughRemoteDevice `json:"devices"`
}

// Only addresses the exporter policy allows are dialled.
type ListPassthroughReq struct {
	Exporter string `validate:"required,hostname_port|hostname|ip"`
}

// Exporter is dialled and never spawned; BusID is matched against the usbip
// busid grammar before it reaches the wire.
type StartPassthroughReq struct {
	Exporter string `validate:"required,hostname_port|hostname|ip"`
	BusID    string `validate:"required,max=31"`
	Mode     string `validate:"omitempty,oneof=hybrid exact"`
}
